File: app/src/state/notifications.ts
import { AppState, Platform } from 'react-native';
import * as Notifications from 'expo-notifications';
import { router } from 'expo-router';
import { getClient } from './store';

/**
 * On-device (local) notification wiring. The server no longer relays through
 * FCM/Expo — instead it broadcasts lightweight `alert` messages over the
 * WebSocket, and the store turns each into a LOCAL notification here. So there
 * is no push credential / EAS setup; the tradeoff is alerts only fire while the
 * app is alive with a live socket (foreground or a short background window).
 *
 * This module owns: foreground display, the Android channel, the "approval"
 * category with Approve / Deny action buttons, presenting local notifications,
 * and the tap / action response handler that resolves prompts over REST (so it
 * works even with no WebSocket connected) or opens the relevant session.
 */

let handlersInstalled = false;
/** The session the user is currently viewing in the foreground (suppresses its own banners). */
let activeSessionId: string | null = null;

// Show notifications even while the app is foregrounded.
Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowBanner: true,
    shouldShowList: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
});

async function ensureChannelAndCategory(): Promise<void> {
  if (Platform.OS === 'android') {
    await Notifications.setNotificationChannelAsync('default', {
      name: 'Default',
      importance: Notifications.AndroidImportance.HIGH,
      vibrationPattern: [0, 200, 100, 200],
    }).catch(() => {});
  }
  await Notifications.setNotificationCategoryAsync('approval', [
    { identifier: 'approve', buttonTitle: 'Approve', options: { opensAppToForeground: false } },
    { identifier: 'deny', buttonTitle: 'Deny', options: { opensAppToForeground: false, isDestructive: true } },
  ]).catch(() => {});
}

/** Record which session is on screen so we can skip notifying about it while foregrounded. */
export function setActiveSession(id: string | null): void {
  activeSessionId = id;
}

/** Request notification permission and set up the Android channel + action
 * category. Returns whether notifications are usable. No push token, no server
 * registration — everything is local now. */
export async function registerForNotifications(): Promise<boolean> {
  try {
    const existing = await Notifications.getPermissionsAsync();
    let status = existing.status;
    if (status !== 'granted') status = (await Notifications.requestPermissionsAsync()).status;
    if (status !== 'granted') return false;
    await ensureChannelAndCategory();
    return true;
  } catch {
    return false;
  }
}

export interface LocalAlert {
  sessionId: string;
  kind: string;
  title: string;
  body: string;
  requestId?: string;
  categoryId?: string;
}

/** Present a local notification for a server `alert`. Suppressed when the user
 * is already looking at that session in the foreground (the in-app sheet/cards
 * already surface it, so a banner would just be noise). */
export async function presentLocalNotification(a: LocalAlert): Promise<void> {
  if (AppState.currentState === 'active' && a.sessionId === activeSessionId) return;
  try {
    await ensureChannelAndCategory();
    await Notifications.scheduleNotificationAsync({
      content: {
        title: a.title,
        body: a.body,
        sound: 'default',
        data: { sessionId: a.sessionId, requestId: a.requestId, kind: a.kind },
        ...(a.categoryId ? { categoryIdentifier: a.categoryId } : {}),
      },
      trigger: null, // deliver immediately
    });
  } catch {
    /* ignore */
  }
}

/** Install the tap / action-button response handler exactly once. */
export function installNotificationHandlers(): void {
  if (handlersInstalled) return;
  handlersInstalled = true;

  Notifications.addNotificationResponseReceivedListener((resp) => {
    const data = resp.notification.request.content.data as {
      sessionId?: string;
      requestId?: string;
      kind?: string;
    };
    const action = resp.actionIdentifier;
    const client = getClient();

    if (data?.kind === 'permission' && data.sessionId && data.requestId && (action === 'approve' || action === 'deny')) {
      client?.respondPermissionRest(data.sessionId, data.requestId, action === 'approve' ? 'allow' : 'deny').catch(() => {});
      return;
    }

    // Any other tap (including the question/done notifications) opens the session.
    if (data?.sessionId) {
      router.push(`/session/${data.sessionId}`);
    }
  });
}

File: flutter-app/lib/state/notifications.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';

/// On-device local notifications (mirrors app/src/state/notifications.ts).
///
/// The server has no FCM/remote push — it broadcasts `alert` messages over the
/// live WebSocket and the client turns each into a local notification. This only
/// fires while the app process is alive and the WS is connected (foreground or a
/// short background window) — the inherent trade-off of dropping FCM.
final FlutterLocalNotificationsPlugin _plugin = FlutterLocalNotificationsPlugin();

/// The session currently on screen — suppress its banners (already being viewed).
String? _activeSessionId;
bool _initialized = false;

/// Callback the app wires up so tapping a notification can route to a session.
void Function(String sessionId)? onNotificationTapSession;

void setActiveSession(String? id) => _activeSessionId = id;

Future<void> initNotifications() async {
  if (_initialized) return;
  const android = AndroidInitializationSettings('@mipmap/ic_launcher');
  const ios = DarwinInitializationSettings();
  await _plugin.initialize(
    const InitializationSettings(android: android, iOS: ios),
    onDidReceiveNotificationResponse: (resp) {
      final sid = resp.payload;
      if (sid != null && sid.isNotEmpty) onNotificationTapSession?.call(sid);
    },
  );

  // Android 13+ runtime permission + channel.
  final androidImpl =
      _plugin.resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>();
  await androidImpl?.requestNotificationsPermission();
  await androidImpl?.createNotificationChannel(const AndroidNotificationChannel(
    'claude-remote',
    'Claude Remote',
    description: 'Approvals, questions, and turn-completion alerts',
    importance: Importance.high,
  ));

  _initialized = true;
}

int _idCounter = 1000;

Future<void> presentLocalNotification({
  required String sessionId,
  required String kind, // permission | question | done
  required String title,
  required String body,
  String? requestId,
  String? categoryId,
}) async {
  // Don't double-buzz for the session already on screen.
  if (sessionId == _activeSessionId) return;
  if (!_initialized) {
    try {
      await initNotifications();
    } catch (e) {
      debugPrint('notif init failed: $e');
      return;
    }
  }

  final android = AndroidNotificationDetails(
    'claude-remote',
    'Claude Remote',
    importance: Importance.high,
    priority: Priority.high,
    category: kind == 'permission' ? AndroidNotificationCategory.call : null,
  );
  final details = NotificationDetails(android: android, iOS: const DarwinNotificationDetails());

  await _plugin.show(_idCounter++, title, body, details, payload: sessionId);
}
